### PR TITLE
mon/OSDMonitor: two "ceph osd crush class rm" fixes

### PR DESCRIPTION
**Pull request by xiexingguo**
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

**Comment by xiexingguo** [comment]
@Songweibin Can you review?

**Comment by xiexingguo** [comment]
retest this please

File: src/mon/OSDMonitor.cc [hunk 1]
```diff
@@ -8610,6 +8609,26 @@ bool OSDMonitor::prepare_command_impl(MonOpRequestRef op,
...
+       for (auto &j: i.second) {
+         if ("crush-device-class" == j.first && device_class == j.second) {
+           referenced_by.push_back(i.first);
+         }
+       }
+     }
+     if (!referenced_by.empty()) {
+       err = -EBUSY;
+       ss << "class '" << device_class
+          << "' is still referenced by erasure-code-profile(s):" << referenced_by;
```
Comment (Songweibin) [inline src/mon/OSDMonitor.cc]:
```suggestion
          << "' is still referenced by erasure-code-profile(s): " << referenced_by;
```
Nit: better to add a space after ":".

**Comment by xiexingguo** [comment]
retest this please

**Comment by xiexingguo** [comment]
@liewegas Mind reviewing this? Thanks

File: src/mon/OSDMonitor.cc [hunk 2]
```diff
@@ -8598,8 +8598,7 @@ bool OSDMonitor::prepare_command_impl(MonOpRequestRef op,
      CrushWrapper newcrush;
      _get_pending_crush(newcrush);
      if (!newcrush.class_exists(device_class)) {
-       err = -ENOENT;
-       ss << "class '" << device_class << "' does not exist";
+       err = 0; // make command idempotent
```
Comment (liewegas) [inline src/mon/OSDMonitor.cc]: This might need to wait for the commit, since we're testing against newcrush (uncommitted leading edge), not crush (committed state).

Comment (liewegas) [reply src/mon/OSDMonitor.cc]: I would test against crush higher up (and goto reply), and then make this test goto update (or retry).

Comment (jecluis) [reply src/mon/OSDMonitor.cc]: I agree, this will need to either wait or check with the stable map as well.

Say we have a given `foo` class in crush; calling
```
ceph osd class rm foo &
ceph osd class rm foo
```
may have us in a situation such that both commands will act during the same proposal, with one removing class `foo` before the other. The second run can only be returned to the user as if the class has been removed IFF we can ensure it didn't exist to begin with, or once we are sure the proposal has been committed. Otherwise we'll be exposing uncommitted state to the user.

**Comment by xiexingguo** [comment]
@liewegas  Updated.

File: src/mon/OSDMonitor.cc [hunk 3]
```diff
@@ -8598,9 +8598,8 @@ bool OSDMonitor::prepare_command_impl(MonOpRequestRef op,
      CrushWrapper newcrush;
      _get_pending_crush(newcrush);
      if (!newcrush.class_exists(device_class)) {
-       err = -ENOENT;
-       ss << "class '" << device_class << "' does not exist";
-       goto reply;
+       err = 0; // make command idempotent
+       goto wait;
```
Comment (liewegas) [inline src/mon/OSDMonitor.cc]: if you goto wait here, you *also* need
```
if (!crush.class_exists(device_class)) {
￼        err = -ENOENT;
￼        ss << "class '" << device_class << "' does not exist";
￼        goto reply;
    }
```
a bit further up

Comment (xiexingguo) [reply src/mon/OSDMonitor.cc]: @liewegas  Do you mean:

```
if (!crush.class_exists(device_class)) {
￼        err = 0;
￼        ss << "class '" << device_class << "' does not exist";
￼        goto reply;
    }
```
?

**Comment by xiexingguo** [comment]
@liewegas Repushed and build now

**Comment by xiexingguo** [comment]
@liewegas Ping?

**Comment by xiexingguo** [comment]
http://pulpito.ceph.com/xxg-2018-10-26_07:06:26-rados-wip-xxg-testing-2018-10-25-0909-distro-basic-smithi/
